test/decl/enum/enumtest.swift

```diff
@@ -441,4 +441,21 @@ class C {}
 protocol P {}
 
 enum E : C & P {}
-// expected-error@-1 {{inheritance from class-constrained protocol composition type 'C & P'}}
\ No newline at end of file
+// expected-error@-1 {{inheritance from class-constrained protocol composition type 'C & P'}}
+
+// SR-11522
+
+enum EnumWithStaticNone1 {
+  case a
+  static let none = 1
+}
+
+enum EnumWithStaticNone2 {
+  case a
+  static let none = EnumWithStaticNone2.a
+}
+
+let _: EnumWithStaticNone1? = .none // Okay
+let _: EnumWithStaticNone2? = .none // expected-warning {{assuming you mean 'Optional<EnumWithStaticNone2>.none'; did you mean 'EnumWithStaticNone2.none' instead?}}
+// expected-note@-1 {{explicitly specify 'Optional' to silence this warning}}{{31-31=Optional}}
+// expected-note@-2 {{use 'EnumWithStaticNone2.none' instead}}{{31-31=EnumWithStaticNone2}}
```

Comment:
How do these test cases behave? How *should* they behave?

```swift
enum GenericEnumWithStaticNone<T> {
  case a
  static var none: GenericEnumWithStaticNone<Int> { .a }
}

let _: GenericEnumWithStaticNone<Int>? = .none
let _: GenericEnumWithStaticNone<String>? = .none
let _: GenericEnumWithStaticNone? = .none
```

Reply:
I don't think this should necessarily block *this* PR, though.

Reply:
Another variation:

```swift
enum GenericEnum<T> {
  case a
}
extension GenericEnum where T == Int {
  static var none: GenericEnumWithStaticNone<Int> { .a }
}
```

Reply:
None of them produce a warning at the moment.

Reply:
I think the first one should produce a warning, but the latter two shouldn't since `none`'s type is `GenericEnumWithStaticNone<Int>` but only first one has it as the base type.

@jrose-apple Did you mean `static var none: GenericEnum <Int> { .a }`?

Reply:
```swift
enum GenericEnumWithStaticNone<T> {
  case a
  static var none: GenericEnumWithStaticNone<Int> { .a }
}

let _: GenericEnumWithStaticNone<Int>? = .none // warning
let _: GenericEnumWithStaticNone<String>? = .none // okay
let _: GenericEnumWithStaticNone? = .none // okay

enum GenericEnum<T> {
  case a
}

extension GenericEnum where T == Int {
  static var none: GenericEnum<Int> { .a }
}

let _: GenericEnum<Int>? = .none // warning
let _: GenericEnum<String>? = .none // okay
```

Reply:
Alright, should be diagnosed now.